sdk/storage/azure-storage-blob/azure/storage/blob/_shared/policies_async.py

```diff
@@ -218,3 +221,24 @@ def get_backoff_time(self, settings):
             if self.backoff > self.random_jitter_range else 0
         random_range_end = self.backoff + self.random_jitter_range
         return random_generator.uniform(random_range_start, random_range_end)
+
+
+class AsyncStorageBearerTokenCredentialPolicy(AsyncBearerTokenCredentialPolicy):
+    """ Custom Bearer token credential policy for following Storage Bearer challenges """
+
+    def __init__(self, credential, **kwargs):
+        # type: (AsyncTokenCredential, **Any) -> None
+        super(AsyncStorageBearerTokenCredentialPolicy, self).__init__(credential, STORAGE_OAUTH_SCOPE, **kwargs)
+
+    async def on_challenge(self, request, response):
+        # type: (PipelineRequest, PipelineResponse) -> bool
+        try:
+            auth_header = response.http_response.headers.get("WWW-Authenticate")
+            challenge = StorageHttpChallenge(auth_header)
+        except ValueError:
+            return False
+
+        scope = challenge.resource_id + DEFAULT_OAUTH_SCOPE
+        await self.authorize_request(request, scope, tenant_id=challenge.tenant_id)
```

Comment:
Since for disk scenario, two requests will be sent:
- first one for storage scope
- second one for disk scope.

There would exist bugs when fetching `data_stream_total`: https://github.com/Azure/azure-sdk-for-python/blob/f424ae46c43dfd92265c55bc81f2df343965d80e/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/policies.py#L306-L313

From the first request, we will get a response without `content-range` as it returns auth failure. The correct `content-range` is returned in second response. But at that time, `data_stream_total` won't change anymore.

This causes `data_stream_current` > `data_stream_total`, and it breaks CLI progress bar.

Reply:
Thanks for finding the issue! I was not aware of this form of progress tracking, and it was indeed broken for Bearer challenges. I have hopefully addressed this in the latest update.